[GRPH-81] Different performance optimization

gladcow opened the pull request:
Backport of the Bitshares PR https://github.com/bitshares/bitshares-core/pull/1180.

It adds several optimization:
 - cache several objects as class members instead of recreation every time we need them;
 - optimize `update_expired_feeds` with tracking of the expired rates;
 - a set of minor optimizations;

bobinson commented on libraries/chain/asset_evaluator.cpp:
```diff
@@ -483,7 +489,21 @@ void_result asset_update_evaluator::do_apply(const asset_update_operation& o)
    }
 
-   d.modify(*asset_to_update, [&](asset_object& a) {
+   // For market-issued assets, if core change rate changed, update flag in bitasset data
```
Not sure whether this is applicable for us. Can you confirm ?

gladcow replied:
One of the optimizations,  introduced in b67f0523, is the tracking of the current exchange rates, instead of iterating of all `asset_object`s for every new block to check if the feed is expired. This optimization works when we have many assets.

gladcow commented:
Also ,  this code doesn't change the functionality itself,  so we can use existing `chain_test` and 'app_test` tests to check if it is correct. I hope it addresses the issues you have pointed on,  so,  re-review,  please,  @bobinson

gladcow commented:
I've updated the state,  re-review,  please.

oxarbitrage commented:
From bitshares PR https://github.com/bitshares/bitshares-core/pull/1180/commits/78ff14edd67462a6b6a9729e4ffc055731d6d67d is not included. Does is not apply for us ?

gladcow commented:
@oxarbitrage , I don't sure that skipping this check is acceptable for Peerplays network, because we haven't details about the workers now.) I've thought that we can add this commit later,  it is not criticaltical. But I can add it if you hink we should apply it right now.

oxarbitrage commented:
I think we should add if we can so we will have the full PR from bitshares applied and will not have to get back later. Thanks.

gladcow commented:
@oxarbitrage , I've added this commit, re-review, please.
